contracts-periphery/src/test/UmbraBatchSend.t.sol

```diff
@@ -0,0 +1,160 @@
+    uint256 totalToll = toll *  sendToken.length;
+    token.approve(address(router), totalAmount);
+
+    vm.expectCall(address(router), abi.encodeWithSelector(router.batchSendTokens.selector, toll, sendToken));
+    vm.expectCall(umbra, abi.encodeWithSelector(IUmbra(umbra).sendToken.selector));
+    router.batchSendTokens{value: totalToll}(toll, sendToken);
+
+    assertEq(token.balanceOf(umbra), umbraPrevBal + totalAmount);
+  }
+
+  function testExpectRevert_BatchSendTokens() public {
+    sendToken.push(UmbraBatchSend.SendToken(alice, tokenAddr, 1 ether, pkx, ciphertext));
+    sendToken.push(UmbraBatchSend.SendToken(bob, tokenAddr, 1 ether, pkx, ciphertext));
+
+    token.approve(address(router), 2 ether);
+    vm.expectRevert(abi.encodeWithSelector(ValueMismatch.selector));
+    router.batchSendTokens{value: toll * 2 + 42}(toll, sendToken);
+  }
+
+  function testFuzz_BatchSend(uint72 amount, uint72 amount2) public {
+    vm.assume(amount > 0);
+    vm.assume(amount2 > 0);
+
+    uint256 totalAmount = uint256(amount) + uint256(amount2);
+
+    sendEth.push(UmbraBatchSend.SendEth(payable(alice), amount, pkx, ciphertext));
+    sendEth.push(UmbraBatchSend.SendEth(payable(bob), amount2, pkx, ciphertext));
+
+    sendToken.push(UmbraBatchSend.SendToken(alice, tokenAddr, amount, pkx, ciphertext));
+    sendToken.push(UmbraBatchSend.SendToken(bob, tokenAddr, amount2, pkx, ciphertext));
+    token.approve(address(router), totalAmount);
+
+    vm.expectCall(address(router), abi.encodeWithSelector(router.batchSend.selector, toll, sendEth, sendToken));        
+    vm.expectCall(umbra, abi.encodeWithSelector(IUmbra(umbra).sendEth.selector));
+    vm.expectCall(umbra, abi.encodeWithSelector(IUmbra(umbra).sendToken.selector));
+
+    uint256 totalToll = toll * sendEth.length + toll * sendToken.length;
+
+    router.batchSend{value: totalAmount + totalToll}(toll, sendEth, sendToken);
+
+    assertEq(alice.balance, alicePrevBal + amount);
+    assertEq(bob.balance, bobPrevBal + amount2);
+    assertEq(token.balanceOf(umbra), umbraPrevBal + totalAmount);
+    }
+
+  function testExpectRevert_BatchSend() public {
+    sendEth.push(UmbraBatchSend.SendEth(payable(alice), 1e16, pkx, ciphertext));
+    sendEth.push(UmbraBatchSend.SendEth(payable(bob), 1e16, pkx, ciphertext));
+
+    sendToken.push(UmbraBatchSend.SendToken(alice, tokenAddr, 1e17, pkx, ciphertext));
+    sendToken.push(UmbraBatchSend.SendToken(bob, tokenAddr, 1e17, pkx, ciphertext));
+    token.approve(address(router), 1e17 * 2);
+
+    uint256 totalToll = toll * sendEth.length + toll * sendToken.length;
+    vm.expectRevert(abi.encodeWithSelector(ValueMismatch.selector));
+    router.batchSend{value: 1e16*2 + totalToll + 42}(toll, sendEth, sendToken);
+  }
+
+}
+
+contract BatchSendWithTollTest is UmbraBatchSendTest {
+  using stdStorage for StdStorage;
+
+  function setUp() public override {
+    super.setUp();
+    stdstore.target(address(umbra)).sig("toll()").checked_write(0.1 ether);
```

Comment:
For readability, I'd add a `setToll(uint256)` method in DSTestPlus that does this. Will likely be useful for future contracts too